### (1) Create a variable that contains the weighted average of the grades in the dataset. The number of students in the fourth grade is different from that of the eighth grade. So you will need a weighted average!
Veri kümesindeki notların ağırlıklı ortalamasını içeren bir değişken oluşturun. Dördüncü sınıftaki öğrenci sayısı sekizinci sınıftaki öğrenci sayısından farklıdır. Yani ağırlıklı bir ortalamaya ihtiyacınız olacak!

In [20]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import warnings

warnings.filterwarnings('ignore')
df = pd.read_csv('states_all.csv')
df.head()

PRIMARY_KEY       STATE  YEAR  ENROLL  TOTAL_REVENUE  FEDERAL_REVENUE  \
0     1992_ALABAMA     ALABAMA  1992     NaN      2678885.0         304177.0   
1      1992_ALASKA      ALASKA  1992     NaN      1049591.0         106780.0   
2     1992_ARIZONA     ARIZONA  1992     NaN      3258079.0         297888.0   
3    1992_ARKANSAS    ARKANSAS  1992     NaN      1711959.0         178571.0   
4  1992_CALIFORNIA  CALIFORNIA  1992     NaN     26260025.0        2072470.0   

   STATE_REVENUE  LOCAL_REVENUE  TOTAL_EXPENDITURE  INSTRUCTION_EXPENDITURE  \
0      1659028.0       715680.0          2653798.0                1481703.0   
1       720711.0       222100.0           972488.0                 498362.0   
2      1369815.0      1590376.0          3401580.0                1435908.0   
3       958785.0       574603.0          1743022.0                 964323.0   
4     16546514.0      7641041.0         27138832.0               14358922.0   

   ...  GRADES_4_G  GRADES_8_G  GRADES_12_G  GRADES_1_8_G  GRADES_9_12_G  \
0  ...     57948.0     58025.0      41167.0           NaN            NaN   
1  ...      9748.0      8789.0       6714.0           NaN            NaN   
2  ...     55433.0     49081.0      37410.0           NaN            NaN   
3  ...     34632.0     36011.0      27651.0           NaN            NaN   
4  ...    418418.0    363296.0     270675.0           NaN            NaN   

   GRADES_ALL_G  AVG_MATH_4_SCORE  AVG_MATH_8_SCORE  AVG_READING_4_SCORE  \
0      731634.0             208.0             252.0                207.0   
1      122487.0               NaN               NaN                  NaN   
2      673477.0             215.0             265.0                209.0   
3      441490.0             210.0             256.0                211.0   
4     5254844.0             208.0             261.0                202.0   

   AVG_READING_8_SCORE  
0                  NaN  
1                  NaN  
2                  NaN  
3                  NaN  
4                  NaN  

[5 rows x 25 columns]

In [37]:
total_4 = (df.AVG_MATH_4_SCORE + df.AVG_READING_4_SCORE)/2
total_8 = (df.AVG_MATH_8_SCORE + df.AVG_READING_8_SCORE)/2
df['total_mean']=total_mean = (total_4 + total_8) /2


In [39]:
df

PRIMARY_KEY          STATE  YEAR  ENROLL  TOTAL_REVENUE  \
0           1992_ALABAMA        ALABAMA  1992     NaN      2678885.0   
1            1992_ALASKA         ALASKA  1992     NaN      1049591.0   
2           1992_ARIZONA        ARIZONA  1992     NaN      3258079.0   
3          1992_ARKANSAS       ARKANSAS  1992     NaN      1711959.0   
4        1992_CALIFORNIA     CALIFORNIA  1992     NaN     26260025.0   
...                  ...            ...   ...     ...            ...   
1710       2019_VIRGINIA       VIRGINIA  2019     NaN            NaN   
1711     2019_WASHINGTON     WASHINGTON  2019     NaN            NaN   
1712  2019_WEST_VIRGINIA  WEST_VIRGINIA  2019     NaN            NaN   
1713      2019_WISCONSIN      WISCONSIN  2019     NaN            NaN   
1714        2019_WYOMING        WYOMING  2019     NaN            NaN   

      FEDERAL_REVENUE  STATE_REVENUE  LOCAL_REVENUE  TOTAL_EXPENDITURE  \
0            304177.0      1659028.0       715680.0          2653798.0   
1            106780.0       720711.0       222100.0           972488.0   
2            297888.0      1369815.0      1590376.0          3401580.0   
3            178571.0       958785.0       574603.0          1743022.0   
4           2072470.0     16546514.0      7641041.0         27138832.0   
...               ...            ...            ...                ...   
1710              NaN            NaN            NaN                NaN   
1711              NaN            NaN            NaN                NaN   
1712              NaN            NaN            NaN                NaN   
1713              NaN            NaN            NaN                NaN   
1714              NaN            NaN            NaN                NaN   

      INSTRUCTION_EXPENDITURE  ...  GRADES_8_G  GRADES_12_G  GRADES_1_8_G  \
0                   1481703.0  ...     58025.0      41167.0           NaN   
1                    498362.0  ...      8789.0       6714.0           NaN   
2                   1435908.0  ...     49081.0      37410.0           NaN   
3                    964323.0  ...     36011.0      27651.0           NaN   
4                  14358922.0  ...    363296.0     270675.0           NaN   
...                       ...  ...         ...          ...           ...   
1710                      NaN  ...         NaN          NaN           NaN   
1711                      NaN  ...         NaN          NaN           NaN   
1712                      NaN  ...         NaN          NaN           NaN   
1713                      NaN  ...         NaN          NaN           NaN   
1714                      NaN  ...         NaN          NaN           NaN   

      GRADES_9_12_G  GRADES_ALL_G  AVG_MATH_4_SCORE  AVG_MATH_8_SCORE  \
0               NaN      731634.0             208.0             252.0   
1               NaN      122487.0               NaN               NaN   
2               NaN      673477.0             215.0             265.0   
3               NaN      441490.0             210.0             256.0   
4               NaN     5254844.0             208.0             261.0   
...             ...           ...               ...               ...   
1710            NaN           NaN             247.0             287.0   
1711            NaN           NaN             240.0             286.0   
1712            NaN           NaN             231.0             272.0   
1713            NaN           NaN             242.0             289.0   
1714            NaN           NaN             246.0             286.0   

      AVG_READING_4_SCORE  AVG_READING_8_SCORE  total_mean  
0                   207.0                  NaN         NaN  
1                     NaN                  NaN         NaN  
2                   209.0                  NaN         NaN  
3                   211.0                  NaN         NaN  
4                   202.0                  NaN         NaN  
...                   ...                  ...         ...  
1710                224.0     

### (2) What is the correlation between the variable you just created and the types of expenditures? Which expenditure item has more correlation than others?
Az önce oluşturduğunuz değişken ile harcama türleri arasındaki ilişki nedir? Hangi harcama kalemi diğerlerinden daha fazla korelasyona sahiptir?

In [48]:
expenditure_df = df.iloc[:,[8,9,10,11,25]]
expenditure_korelasyon = expenditure_df.corr()
expenditure_korelasyon

TOTAL_EXPENDITURE  INSTRUCTION_EXPENDITURE  \
TOTAL_EXPENDITURE                      1.000000                 0.991334   
INSTRUCTION_EXPENDITURE                0.991334                 1.000000   
SUPPORT_SERVICES_EXPENDITURE           0.993309                 0.976857   
OTHER_EXPENDITURE                      0.946084                 0.909262   
total_mean                             0.019821                 0.038317   

                              SUPPORT_SERVICES_EXPENDITURE  OTHER_EXPENDITURE  \
TOTAL_EXPENDITURE                                 0.993309           0.946084   
INSTRUCTION_EXPENDITURE                           0.976857           0.909262   
SUPPORT_SERVICES_EXPENDITURE                      1.000000           0.953109   
OTHER_EXPENDITURE                                 0.953109           1.000000   
total_mean                                        0.023362          -0.045637   

                              total_mean  
TOTAL_EXPENDITURE               0.019821  
INSTRUCTION_EXPENDITURE         0.038317  
SUPPORT_SERVICES_EXPENDITURE    0.023362  
OTHER_EXPENDITURE              -0.045637  
total_mean                      1.000000

total_mean hepsiyle çok düşük korelasyona sahip. total_mean ile en yüksek korelasyona sahip olan INSTRUCTION_EXPENDITURE.

### (3) Now apply the Principal Components Analysis (PCA) for the four expenditure items! How much of the total variance can be explained by the first component?
Şimdi dört harcama kalemi için Temel Bileşenler Analizini (PCA) uygulayın! Toplam varyansın ne kadarı birinci bileşenle açıklanabilir?

In [52]:
expenditure_items = df.iloc[:,[8,9,10,11]]
expenditure_items.dropna(inplace=True)
expenditure_items

TOTAL_EXPENDITURE  INSTRUCTION_EXPENDITURE  \
51            2833433.0                1564558.0   
52            1126398.0                 494917.0   
53            3623946.0                1578889.0   
54            1376067.0                 782791.0   
55           28110986.0               15281147.0   
...                 ...                      ...   
1270         16497520.0                8944614.0   
1271         15253296.0                7087365.0   
1272          3366566.0                1804235.0   
1273         11787535.0                5760418.0   
1274          2034229.0                 921494.0   

      SUPPORT_SERVICES_EXPENDITURE  OTHER_EXPENDITURE  
51                        794146.0           237222.0  
52                        433788.0            36291.0  
53                       1000914.0           164083.0  
54                        386526.0            68617.0  
55                       8914559.0          1608514.0  
...                            ...                ...  
1270                     5164699.0           643506.0  
1271                     4904575.0           577606.0  
1272                     1127255.0           236487.0  
1273                     3691439.0           508013.0  
1274                      585700.0            53570.0  

[1224 rows x 4 columns]

In [54]:
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

X=expenditure_items.values

X = StandardScaler().fit_transform(expenditure_items)
pca = PCA(n_components=4)
principalComponents=pca.fit_transform(X)
exp_var= pca.explained_variance_ratio_
cumsum_var=np.cumsum(np.round(pca.explained_variance_ratio_, decimals=4)*100)
print(exp_var)
print(cumsum_var)

[9.71355694e-01 2.46874518e-02 3.49316153e-03 4.63692535e-04]
[ 97.14  99.61  99.96 100.01]


Birinci bileşenle %97.14'ü açıklanabiliyor. Bu harika bir oran. Diğerleri çok az açıklama oranına sahipler.

### (4) What is the correlation between the GPA you created and the first principal component?
Oluşturduğunuz GPA ile ilk temel bileşen arasındaki ilişki nedir?

In [55]:
expenditure_gpa = df.iloc[:,[8,25]]
expenditure_gpa_korelasyon = expenditure_gpa.corr()
expenditure_gpa_korelasyon

TOTAL_EXPENDITURE  total_mean
TOTAL_EXPENDITURE           1.000000    0.019821
total_mean                  0.019821    1.000000

Pozitif bir korelasyona sahiptir ancak çok yüksek değildir bu değer.

### (5) When you need to choose the most appropriate variables for your model, would you prefer the first basic variables instead of the expenditure items? Why?
Modelinize en uygun değişkenleri seçmeniz gerektiğinde harcama kalemleri yerine ilk temel değişkenleri mi tercih edersiniz? Niye

Cevap: PC1 olan TOTAL_EXPENDITURE'i seçerim çünkü zaten tek başına açıklayıcılığı yüzde 97 den fazla. Diğerlerinin açıklayıcılığı çok düşük olduğu için onları çıkarabiliriz.